docs/decisions/0017-reimplement-asset-processing.rst

```diff
@@ -0,0 +1,268 @@
+
+All of these potential solutions would involve refactoring or entirely replacing parts of the current asset processing system.
+
+
+Decision
+********
+
+We will largely reimplement edx-platform's asset processing system. We will aim to:
+
+* Use well-known, npm-installed frontend tooling wherever possible.
+* When bespoke processing is required, use standard POSIX tools like Bash.
+* When Django/Python is absolutely required, contain its impact so that the rest of the system remains Python-free.
+* Avoid unnecessary indirection or abstraction. For this task, extensibility is a non-goal, and simplicity is a virtue.
+* Provide a clear migration path from the old system to the new one.
+* Enable the future removal of as much legacy frontend tooling code as possible.
+
+Consequences
+************
+
+Reimplementation Specification
+==============================
+
+The three top-level edx-platform asset processing actions are *build*, *collect*, and *watch*. The build action can be further broken down into five stages. Here is how those actions and stages will be reimplemented:
+
+
+.. list-table::
+   :header-rows: 1
+
+   * - Description
+     - Old implementation
+     - New implementation
+
+   * - **Build: All stages.** Compile, generate, copy, and otherwise process static assets so that they can be used by the Django webserver or collected elsewhere. For many Web applications, all static asset building would be coordinated via Webpack or another NPM-managed tool. Due to the age of edx-platform and its legacy XModule and Comprehensive Theming systems, though, there are five stages which need to be performed in a particular order.
+
+     - ``paver update_assets --skip-collect``
+
+       A Python-defined task that calls out to each build stage.
+
+     - ``assets/build.sh``
+
+       A Bash script that contains all build stages, with subcommands available for running each stage separately. Its command-line interface inspired by Tutor's ``openedx-assets`` script. The script will be runnable on any POSIX system, including macOS and Ubuntu and it will linted for common shell scripting mistakes using `shellcheck <https://www.shellcheck.net>`_.
+     
+   * - + **Build stage 1: Copy npm-installed assets** from node_modules to other folders in edx-platform. They are used by certain especially-old legacy LMS & CMS frontends that are not set up to work with npm directly.
+
+     - ``paver update_assets --skip-collect``
+
+       Implemented in Python within update_assets. There is no standalone command for it.
+
+     - ``assets/build.sh npm``
+
+       Pure Bash reimplementation.
+ 
+   * - + **Build stage 2: Copy XModule fragments** from the xmodule source tree over to input directories for Webpack and SCSS compilation. This is required for a hard-coded list of old XModule-style XBlocks. This is not required for new pure XBlocks, which include (or pip-install) their assets into edx-platform as ready-to-serve JS/CSS/etc fragments.
```

Comment:
@feanil Great question. [This discovery doc](https://docs.google.com/document/d/1FqsvXpvrzsi2Ekk9RttUpcT2Eg0NxenFmV52US_psFU/edit#heading=h.ogslfhqqejyh) just dropped; I'm working through it now in hopes that it'll lead me to an answer. I'll update the text here when I'm done.

Reply:
Also related: https://github.com/openedx/edx-platform/issues/31624

Reply:
It's temporary, but it might be one of those temporary-forever problems. Hard to tell right now.

I updated this step to make it clearer what is the immediate solution (use xmodule_assets), the next step (rewrite xmodule_assets in Bash), and then the ideal final state (make them normal xblocks).